### ISSUE TITLE
Slow compilation and optimization of a big Theano graph

### ISSUE THREAD

**Issue by MarcCote** [Theano/Theano#4494]
Hi, I'm trying to implement something similar to a ResNet (https://github.com/KaimingHe/resnet-1k-layers) in Theano. I based my code this [Lasagne recipe](https://github.com/Lasagne/Recipes/blob/master/papers/deep_residual_learning/Deep_Residual_Learning_CIFAR-10.py).

I'm able to compile and optimize when I use a depth of ~38 layers in about 40 min on the Helios cluster. When I try with 164 layers (which is still far from the 1001 layers reported in the paper), my job stopped (after 12 hours) before the compilation/optimization was finished.

I profile the execution using different options: mode={FAST_COMPILE, FAST_RUN} and optimizer={fast_compile, fast_run} with my model having either 38 or 164 layers. 

Here are the profiling logs:
https://www.dropbox.com/sh/i75qoemmuf8whiw/AADnr2Eq7POpgEjpFAnOEmFia?dl=0

*Don't mind the memory errors as they happen after the compilation. My main concern is the compilation/optimization time.

**Comment by nouiz** [comment]
Thanks for reporting and investigating.

If this happen again, next time you could just send me the profile with
optimizer=fast_compile and the default one. Changing the mode isn't very
useful for this type of investigation.

Can you send me the code so that I can test a few things?

A few notes, fast_compile is faster but not as much. The problem is that
the graph is much bigger and this make the move to the GPU much slower. It
is the bottleneck (1/3 of the compilation time).

Using the default mode with the following optimization flag should generate
a case that is even faster to compile then fast_compile:

optimizer_excluding=inplace_elemwise_optimizer

Was the C code cache full before doing those profile? There is a big part
of the time in the c compilation, so I suppose so.

In addition this PR should help a little:

https://github.com/Theano/Theano/pull/4496


I have a few local modification that I

**Comment by nouiz** [comment]
I updated that PR again. It could help a little bit the fast_compile, but
should help more the fast_run case.

I'm still interrested to get access to the code to continue. You can still
try the flag I suggested in the mean time.

**Comment by lamblin** [comment]
is `inplace_elemwise_optimizer` included in `optimizer=fast_compile`? I don't think it should be the case.

**Comment by nouiz** [comment]
No. It isn't.

But fast_compile make the graph much bigger then fast_run. Like 28k node
intead of 15k. This make the opt that move nodes to the GPU much slower
(800s vs 300s). So most of the speed up from fast_compile is lost. We
probably need to find the few optimizations that make a big difference.

With the timming I see, fast_run without that opt would be faster then
fast_compile. I recall having seen that in the past.

**Comment by MarcCote** [comment]
Thanks @nouiz for your suggestions I will try them. However with NIPS's deadline coming really fast, I don't have a lot of time to investigate this. I'll see if I can run them tonight. The code is currently in a private repository but I'll give you access to it.

**Comment by nouiz** [comment]
I'm running the code. After some discussion with @lamblin, I did a mistake
and using fast_run with the other optimization disable won't give you
something faster then fast_compile.

thanks

**Comment by lachlants** [comment]
I tried compiling a very shallow resnet (20 layer) and the number of graph nodes (before merge1) explodes from ~15K to ~50K when batch normalization is included. I think it has something to do with the derived gradient expression. 

As a workaround I tried using this PR: https://github.com/Theano/Theano/pull/4237

However the code still seems a little buggy, I get an error about the gradient expression not returning the right number of outputs.

**Comment by nouiz** [comment]
Thanks for the information. How did you implement BN? Did you used:

http://deeplearning.net/software/theano/library/tensor/nnet/bn.html#theano.tensor.nnet.bn.batch_normalization

If not, using this should speed up compilation a lot. It use much less node
in the graph.

I did a PR with some updates:

https://github.com/Theano/Theano/pull/4524

With it, you can speed up compilation with optimizer=fast_compile by 5%
when using the flag:

optimizer_including=UnShapeOpt

If you uncomment this change:

https://github.com/Theano/Theano/pull/4524/commits/b3dbc8ba965f4a8d287ae4d0218620a76a3721f0

It make the compilation with those 2 flags even faster, but it make the
final graph bigger. I don't know why.

Most of the time is spend in moving the graph to the GPU when using
optimizer=fast_compile. @sentient07 have already planned to speed up that
part.

I

**Comment by lachlants** [comment]
> Thanks for the information. How did you implement BN? Did you used:

> http://deeplearning.net/software/theano/library/tensor/nnet/bn.html#theano.tensor.nnet.bn.batch_normalization

Yep, I used that op. Improved things a little. 

Hmmm... when I profile the output I see the vast majority of time being spent in the merge1 optimizer (seems to grow non-linearly with the number of input nodes). I think this is because the gradient for batch normalization is highly non-trivial and requires many nodes.

Anyway thanks for your help.

**Comment by nouiz** [comment]
Can you show me the full profile? I know how to help the merge1, but I
never saw it in real life. I wasn't sure if this was happening in practice
or not.

**Comment by Sentient07** [comment]
@MarcCote Thank you for sharing the profiling logs. I would be working on this and this will be really helpful. 😄

**Comment by lachlants** [comment]
@nouiz Here's the profile for my code. I used 0.8.0rc1 since I was getting an error with master in the profile printing (list index out of bounds). 

Resnet (22 conv layers) = 63K nodes, 36 sec merge1: [theano_profile_resnet-20.txt](https://github.com/Theano/Theano/files/273699/theano_profile_resnet-20.txt)
Resnet (34 conv layers) = 156K nodes, 185 sec merge1: [theano_profile_resnet-34.txt](https://github.com/Theano/Theano/files/273711/theano_profile_resnet-34.txt)
Resnet (46 conv layers) = 290K nodes, 677 sec merge1: [theano_profile_resnet-46.txt](https://github.com/Theano/Theano/files/273732/theano_profile_resnet-46.txt)

**Comment by MarcCote** [comment]
@Sentient07 awesome. Let me know if you need something else.

**Comment by nouiz** [comment]
@lachlants, can you give me your code? It would help doing the fix I have
in mind.

The full solution is to make a pass on all Op.grad() method, but with your
code I'll be able to find the one that cause you problem, so it will be
much faster to implement.

**Comment by nouiz** [comment]
Just a note, following discussion on the mailing list, I made a PR that should speed up compilation if using theano bn functions:

https://github.com/Theano/Theano/pull/4543

I don't know why lasagne don't use it.

**Comment by lachlants** [comment]
@nouiz I've made a simple example which demonstrates the issue (it would have been a complex process identifying which parts of my code can be made public): [RESNET-34](https://gist.github.com/lachlants/22a8d1f2ef9416bfc959008c03f0ae63)

The example demonstrates the training graph for a resnet-34 with momentum SGD. Removing batch normalization makes it go from ~90K to ~5K nodes.

**Comment by nouiz** [comment]
I got some speed up in this PR related to your specific use case:

https://github.com/Theano/Theano/pull/4543

I also did one change to your code that remove some duplicated code. I
moved outside the loop the call to theano.grad(). It make this call faster
and remove some merge of nodes:
    gradients = tensor.grad(cost, params)
    for p, g in zip(params, gradients):
        m = theano.shared(numpy.asarray(p.get_value(borrow=True)*0.0,
                                        dtype=theano.config.floatX),
                          broadcastable=p.broadcastable, borrow=True)
        gradient = g + decay * p
        m_update = momentum*m + (1.0 - momentum)*gradient
        p_update = p - lr*m_update
        updates += [(p, p_update), (m, m_update)]

With just that (without the PR), I get from 54s to 34s for the compilation
time. The timing is a little bit strange. But with that change, it remove
the need for the new change I did in 4543 for your specific use case.

So the conclusion, always call grad only once!

That branch still contain speed up when there is more then 1 layers.

**Comment by lachlants** [comment]
Cool thanks. I didn't know the grad() method could take a list of parameters. In my own code I've somewhat side stepped this problem by fixing the cudnn batch normalization PR...

**Comment by nouiz** [comment]
Can you make this change available? Other people are interrested in that.

**Comment by f0k** [comment]
> Just a note, following discussion on the mailing list, I made a PR that should speed up compilation if using theano bn functions:
>
> #4543
>
> I don't know why lasagne don't use it.

Because we wanted the BatchNormLayer to be compatible with cuDNN v4, which stores an exponential moving average of the inverse standard deviation instead of the original variance. We wanted BatchNormLayer and DNNBatchNormLayer to be freely exchangeable. So `T.nnet.batch_norm` didn't fit. See https://github.com/Lasagne/Lasagne/pull/467#issuecomment-170065285 for details.
Unluckily, cuDNN v5 changed the API to compute the variance, so our compatibility efforts were for nothing and we have to either break backwards compatibility or support both cases now.

**Comment by nouiz** [comment]
https://github.com/Theano/Theano/pull/4582 merge the cuDNN batch norm. If you use it, it should speed up compilation significatively.

**Comment by nouiz** [comment]
We just merge a PR that speed up the move of the graph to the GPU in the new back-end. We didn't ported the cuDNN batch norm to the new back-end. So you have 2 options to speed up the compilation for now.
